pkg/registration/hub/gc/controller.go

```diff
@@ -0,0 +1,108 @@
+package gc
+
+import (
+	"context"
+	"time"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	rbacv1client "k8s.io/client-go/kubernetes/typed/rbac/v1"
+	rbacv1listers "k8s.io/client-go/listers/rbac/v1"
+	"k8s.io/client-go/tools/cache"
+
+	informerv1 "open-cluster-management.io/api/client/cluster/informers/externalversions/cluster/v1"
+	clusterv1listers "open-cluster-management.io/api/client/cluster/listers/cluster/v1"
+	worklister "open-cluster-management.io/api/client/work/listers/work/v1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+type gcReconcileOp string
+
+const (
+	gcReconcileRequeue  gcReconcileOp = "gcReconcileRequeue"
+	gcReconcileStop     gcReconcileOp = "gcReconcileStop"
+	gcReconcileContinue gcReconcileOp = "gcReconcileContinue"
+)
+
+// gcReconciler is an interface for reconcile cleanup logic after cluster is deleted.
+// clusterName is from the queueKey, cluster may be nil.
+type gcReconciler interface {
+	reconcile(ctx context.Context, clusterName string, cluster *clusterv1.ManagedCluster) (gcReconcileOp, error)
+}
+
+type GCController struct {
+	clusterLister clusterv1listers.ManagedClusterLister
+	gcReconcilers []gcReconciler
+}
+
+// NewGCController ensures the related resources are cleaned up after cluster is deleted
+func NewGCController(
+	clusterRoleLister rbacv1listers.ClusterRoleLister,
+	roleBindingLister rbacv1listers.RoleBindingLister,
+	namespaceInformer corev1informers.NamespaceInformer,
+	clusterInformer informerv1.ManagedClusterInformer,
+	manifestWorkLister worklister.ManifestWorkLister,
+	rbacClient rbacv1client.RbacV1Interface,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	controller := &GCController{
+		clusterLister: clusterInformer.Lister(),
+		gcReconcilers: []gcReconciler{
+			newGCRoleBindingController(roleBindingLister, namespaceInformer, manifestWorkLister, rbacClient, eventRecorder),
+			newGCClusterRoleController(clusterRoleLister, clusterInformer, manifestWorkLister, rbacClient, eventRecorder),
+		},
+	}
+	return factory.New().
+		WithInformersQueueKeysFunc(func(obj runtime.Object) []string {
+			accessor, _ := meta.Accessor(obj)
+			if accessor.GetDeletionTimestamp() != nil && !accessor.GetDeletionTimestamp().IsZero() {
+				return []string{accessor.GetName()}
+			}
+			return []string{}
+		}, clusterInformer.Informer(), namespaceInformer.Informer()).
```

Comment:
why namespaceInformer here?

Reply:
there is a case that the cluster ns is deleted by user, the work rolebinding in the cluster ns cannot be deleted since there is a finalizer on it. so need to remove the finalizer to make the cluster ns deleted and the cluster ns will be re-created. 
I think we can add a TODO here , will remove namespaceInformer here and to prohibit deleting cluster ns in registration webhook